### PR TITLE
Fix StackOverflow in non-recursive bindings checker

### LINKED ISSUE #12322: 64-bit fscAnyCpu.exe uses more stack than 32-bit fsc.exe (was: Project that builds in VS2019 fails to build in VS2022)
A project that is part of a reasonably complex solution is failing to build in 2022 due to the compiler crashing. I've managed to boil things down to a simple, standalone repro shown below.

**Repro steps**

The following simplified code typifies that which is causing the compiler to crash:

```fsharp
module repro_fscanycpu_failure

//
// This builds fine in VS2019, but fails under VS2022 with:
//
// 1>  Process is terminated due to StackOverflowException.
// 1>  C:\Program Files\Microsoft Visual Studio\2022\Preview\Common7\IDE\CommonExtensions\Microsoft\FSharp\Tools\Microsoft.FSharp.Targets(317,9): error MSB6006: "fscAnyCpu.exe" exited with code -1073741571.
//
// Opening this file in the VS2022 editor will also crash, presumably for the same underlying reason.
//

type ReproBuilder () =
    member _.Delay x = x ()
    member _.Yield  (x) = x
    member _.Combine (x, y) = x + y

let repro = ReproBuilder ()

let callSite () =
    repro {
        // Commenting out some of the below is enough to avoid StackOverflow on my machine.
        0
        1
        2
        3
        4
        5
        6
        7
        8
        9
        0
        1
        2
        3
        4
        5
        6
        7
        8
        9
        0
        1
        2
        3
        4
        5
        6
        7
        8
        9
        0
        1
        2
        3
        4
        5
        6
        7
        8
        9
        0
        1
        2
        3
        4
    }
```

**Expected behavior**

It should build in 2022 as it does in 2019.

**Actual behavior**

It fails with:

> 1>C:\Program Files\Microsoft Visual Studio\2022\Preview\Common7\IDE\CommonExtensions\Microsoft\FSharp\Tools\Microsoft.FSharp.Targets(317,9): error MSB6006: "fscAnyCpu.exe" exited with code -1073741571.

**Known workarounds**

1. **Short-term fix**: Use VS2019.

2. **Recommended**: Make your inputs smaller, probably half the size will do, divide your large expressions into two

**Related information**

Provide any related information (optional):

* Windows 10
* .NET Core

### LINKED ISSUE #16867: Error MSB6006: "fscAnyCpu.exe" exited with code -1073741571
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/Error-MSB6006:-fscAnyCpuexe-exited-wi/10611054)._

---
\[severity:I’m unable to use this version\] \[regression\] \[worked-in:17.8.7\]  
\[error\]C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\Common7\\IDE\\CommonExtensions\\Microsoft\\FSharp\\Tools\\Microsoft.FSharp.Targets(331,9): Error MSB6006: “fscAnyCpu.exe” exited with code -1073741571.  
Visual Studio 2022 build 17.9.2  
Does not happen with build 17.8.7

We have a solution consisting of over 240 projects - most are written in C#, but some (approximately 20) are written in F#.  
The automated CI/CD build ran successfully on 20th Feb against Windows image 20240211.1.0, but since then has failed.  
It always fails with the above error in **2** of the F# projects (the rest of them are fine, except where they depend on the 2 failed projects).  
It fails when I re-run the last successful build.
There have been no changes to the source code, build pipelines or scripts.

I also tried with a hosted agent where the local computer is running VS 2022 Pro build 17.9.2 - same error.

I am able to build the solution on my local computer running VS 2022 Pro build 17.8.7, but not with the automated CI/CD pipelines.

We are now unable to create product releases for our Cloud-based solution and deploy any fixes or changes.

---
#### Feedback Bot on 3/8/2024, 04:16 AM: 

(private comment, text removed)

---
### Original Solutions
(no solutions)

### PR DESCRIPTION
**Pull request by vzarytovskii**
## Description

Fixes #16867 by making main recursive checking function non-cancellable (and tail recursive). It's hard to make builder work under stack guards, since it needs to be calling guard from `Bind`, `ReturnFrom` and `BindReturn`.

This also adds additional activity logging for the stack guards (only when subscribed).

@dotnet/fsharp-team-msft do we want to backport it to VS17.9?

## Checklist

- [ ] ~Test cases added~ not applicable
- [ ] ~Performance benchmarks added in case of performance changes~ not applicable
- [x] Release notes entry updated

### CODE REVIEW

**Review by brianrourkeboll — COMMENTED** [review]
> * ~Test cases added~ not applicable

Would a test that type-checks a module with a few thousand `let`-bindings not make sense?

**Comment by vzarytovskii** [comment]
> > * ~Test cases added~ not applicable
> 
> 
> 
> Would a test that type-checks a module with a few thousand `let`-bindings not make sense?

We have no way of running fscAnyCpu outside of fsharpqa, and I don't want to add it there since we try to get rid of it.

File: src/Compiler/Checking/CheckDeclarations.fs [hunk 1]
```diff
@@ -5470,20 +5478,25 @@ and TcModuleOrNamespaceElements cenv parent endm env xml mutRecNSInfo openDecls0
...
-        let! compiledDefs, envAtEnd = TcModuleOrNamespaceElementsNonMutRec cenv parent typeNames endm ([], env, env) synModuleDecls 
-
-        // Apply the functions for each declaration to build the overall expression-builder 
-        let moduleDefs = List.collect p13 compiledDefs
-        let moduleDefs = match openDecls0 with [] -> moduleDefs | _ -> TMDefOpens openDecls0 :: moduleDefs
-        let moduleContents = TMDefs moduleDefs 
+    | None ->
+        let! ct = Cancellable.token ()
+        let result = TcModuleOrNamespaceElementsNonMutRec cenv parent typeNames endm ([], env, env) synModuleDecls ct
+        //let! compiledDefs, envAtEnd = TcModuleOrNamespaceElementsNonMutRec cenv parent typeNames endm ([], env, env) synModuleDecls ct
```
Comment (KevinRansom) [inline src/Compiler/Checking/CheckDeclarations.fs:5484]: commented code

**Review by KevinRansom — APPROVED** [review]

**Review by psfinaki — APPROVED** [review]

File: src/Compiler/Checking/CheckDeclarations.fs (hunk 1 above)
Comment (vzarytovskii) [inline src/Compiler/Checking/CheckDeclarations.fs:5484]:
```suggestion
```

**Review by T-Gro — APPROVED** [review]

**Comment by T-Gro** [comment]
The description is a little misleading - the behavior is still cancellable, it just does not use the cancellable-builder/CE anymore, right?

**Comment by vzarytovskii** [comment]
> The description is a little misleading - the behavior is still cancellable, it just does not use the cancellable-builder/CE anymore, right?

Well, yes and no. Before it could cancel on more occasions. Now it's only explicit in the beginning + when running cancellable inside.

**Pull request merged**